### PR TITLE
En 10743 acknowledge sent data from covalent

### PR DESCRIPTION
**Pull request by mariusmihaic**
`Covalent indexer` :
- Added `mutex` and `chan` for each type of websocket connection (send/receive). Whenever a new connection is set from outside(different go routine), before overwriting(and closing) the existing one, they will be locked(using the `mutex`) and signaled (using the `chan`)
- Added `SetWSSender` and `SetWSReceiver`, which sets from outside new send/receive ws using the above mentioned mutexes and channels
- Added `getWSS` and `getWSR` which returns existing ws connections safely (using mutex)
- Added `waitForWSSConnection` and `waitForWSRConnection` which wait in a loop until a new connection is established, by reading from the channels and waiting for them to be written
- Added `sendWithRetrial` which **initially - when starting the node**, it should not process anything until both send/receive sockets are connected. Afterwards, every `RetrialTimeoutMS` tick it tries to send data with acknowledge. If successful, it will exit the function, otherwise it will keep trying to send data in an infinite loop. 
- Added `sendDataWithAcknowledge` which will send block data and wait for its block header hash to confirm that the data has been received successfully. In case of an error(either send or receive), it will wait for a new ws connection

`Unit tests`:
- Added various unit tests. However, functionality has also been tested in an "integration test", using a local testnet and clients (which is not visible in this PR)

`Covalent indexer factory`:
- Added handler function for `RouteAcknowledgeData`,  which upgrades a http connection to a websocket connection and sets it to covalent indexer.`

**Deleted**:
- WSSender and every file related to it. Instead of using a struct to send/receive data, an interface is used now (see `process.WSConn`)

### CODE REVIEW

File: covalent_indexer.go [hunk 1]
```diff
@@ -31,91 +39,196 @@ func NewCovalentDataIndexer(processor DataHandler, server *http.Server) (*covale
...
-	c.wss = wss
+	ci.wsr = wsr
+	ci.mutWSR.Unlock()
+
+	ci.newConnectionWSR <- struct{}{}
+}
+
+func (ci *covalentIndexer) getWSS() process.WSConn {
+	ci.mutWSS.RLock()
+	wss := ci.wss
```
Comment (miiu96) [inline covalent_indexer.go]: this is not a copy. you just pass the pointer

Comment (miiu96) [reply covalent_indexer.go]: is the same thing with `return ci.wss`

File: covalent_indexer.go [hunk 2]
```diff
@@ -31,91 +39,196 @@ func NewCovalentDataIndexer(processor DataHandler, server *http.Server) (*covale
...
+	ci.mutWSS.RUnlock()
+
+	return wss
 }
 
-func (c *covalentIndexer) start() {
-	err := c.server.ListenAndServe()
+func (ci *covalentIndexer) getWSR() process.WSConn {
+	ci.mutWSR.RLock()
+	wsr := ci.wsr
```
Comment (miiu96) [inline covalent_indexer.go]: also here

File: covalent_indexer.go [hunk 3]
```diff
@@ -31,91 +39,196 @@ func NewCovalentDataIndexer(processor DataHandler, server *http.Server) (*covale
...
 // SaveValidatorsRating DUMMY
-func (c covalentIndexer) SaveValidatorsRating(indexID string, infoRating []*indexer.ValidatorRatingInfo) {
+func (ci *covalentIndexer) SaveValidatorsRating(string, []*indexer.ValidatorRatingInfo) {
 }
 
 // SaveAccounts DUMMY
-func (c covalentIndexer) SaveAccounts(blockTimestamp uint64, acc []data.UserAccountHandler) {}
+func (ci *covalentIndexer) SaveAccounts(uint64, []data.UserAccountHandler) {}
 
 // Close DUMMY
```
Comment (AdoAdoAdo) [inline covalent_indexer.go]: update comment

File: covalent_indexer_test.go [hunk 4]
```diff
@@ -50,110 +51,350 @@ func TestCovalentIndexer_SetWSSender_SetTwoConsecutiveWebSockets_ExpectFirstOneC
...
-				called.Set()
-				return nil
-			},
+			Addr: "localhost:21119",
+		},
+	)
+	defer ci.Close()
+
+	called1 := atomic.Flag{}
+	called1.Unset()
```
Comment (AdoAdoAdo) [inline covalent_indexer_test.go]: you can remove the unset after init, as it is already unset.

File: covalent_indexer.go (hunk 1 above)
Comment (mariusmihaic) [reply covalent_indexer.go]: Yes, this was intended. Changed the code  to highlight that I don't want to copy the value. See comment in the previous PR

File: covalent_indexer.go (hunk 2 above)
Comment (mariusmihaic) [reply covalent_indexer.go]: Same comment as above. Refactor this get as well

File: covalent_indexer.go (hunk 3 above)
Comment (mariusmihaic) [reply covalent_indexer.go]: Done in previous PR, merged the changes in this as well

File: covalent_indexer_test.go (hunk 4 above)
Comment (mariusmihaic) [reply covalent_indexer_test.go]: Removed unset here and everywhere else in tests

**Review by AdoAdoAdo — APPROVED** [review]

**Review by miiu96 — APPROVED** [review]

**Pull request merged**